src/index.ts

```diff
@@ -8,44 +8,63 @@ import {
   CommentBlock,
 } from "babel-types"
 
+const createCommentBlock = (value: string): Comment => ({
+  type: "CommentBlock",
+  value,
+}) as CommentBlock
+
 const PURE_ANNOTATION = "#__PURE__"
 
-const isPureAnnotated = (comments?: Comment[]): boolean => {
-  if (typeof comments === "undefined") {
+const isPureAnnotated = (nodePath: NodePath<Node>): boolean => {
+  const { leadingComments } = nodePath.node
+
+  if (leadingComments === undefined) {
     return false
   }
-  if (comments.length > 0) {
-    return comments[comments.length - 1].value === PURE_ANNOTATION
+
+  return /[@#]__PURE__/.test(leadingComments[leadingComments.length - 1].value)
+}
+
+const annotateAsPure = (nodePath: NodePath<Node>) => {
+  if (isPureAnnotated(nodePath)) {
+    return;
   }
-  return false
+
+  const node = nodePath.node;
+  node.leadingComments = (node.leadingComments || []).concat(createCommentBlock(PURE_ANNOTATION))
 }
 
-const createComponentBlock = (value: string): Comment => ({
-  type: "CommentBlock",
-  value,
-}) as CommentBlock
+const isTopLevel = (nodePath: NodePath<Node>): boolean => {
+  return nodePath.getStatementParent().parentPath.isProgram()
+}
 
 export default () => ({
   inherits: syntax,
   visitor: {
-    CallExpression: {
-      enter(nodePath: NodePath<Node>) {
-        if (nodePath.parentPath.isVariableDeclarator()
-          || nodePath.parentPath.isAssignmentExpression()
-          || (nodePath.parentPath.isCallExpression()) && ((nodePath.parentPath.get(
```

Comment:
This is needed.
For `fn(fn())` => `fn(/*#__PURE__*/fn())`

Reply:
`fn1(fn2())` if we do not annotate `fn1` (and we shouldnt - no assignment context means side effects) uglify wont be able to get rid of `fn2` call (and it shouldnt be, its result is used after all)

the whole point of the `/*#__PURE__*/` is to be able dead code eliminate unused results

Reply:
Continue https://github.com/morlay/babel-plugin-annotate-pure-call-in-variable-declarator/pull/1#discussion_r140674943

The conditions here should cover all call expressions which are used.

the pure annotated call in used, it will not be removed.
but if it is in some unused chain,  we need to drop it.